File: nucleo/gih_nucleo/guloso.py
"""Os planos gulosos: os mínimos das cotas, e depois o que parece melhor a cada passo.

Dois critérios, porque cada um erra onde o outro acerta:

- **Por razão ganho/custo** (`RAZAO`), que é o certo quando o orçamento aperta.
  Quando quem aperta é o máximo de ações K, ele gasta as vagas em ações baratas
  e deixa orçamento sobrando.
- **Por ganho** (`GANHO`), que é o certo quando K aperta, e erra quando o
  orçamento acaba em poucas ações caras.

Servem a duas coisas:

- **Soluções iniciais do genético.** Viáveis por construção, elas garantem, com
  o elitismo, que o genético nunca devolve algo pior que elas nem inviável
  (ADR-011).
- **Referência de qualidade.** É o que uma planilha faria, e o genético precisa
  superá-lo onde ele erra (H48). O erro é ser míope: escolhe um par por vez e
  não volta atrás.
"""
from __future__ import annotations

from functools import cmp_to_key

from gih_nucleo.problema import Instancia
from gih_nucleo.viabilidade import conjunto_minimo

RAZAO = "razao"
GANHO = "ganho"


def completar(inst: Instancia, genes: list[int], criterio: str = RAZAO) -> list[int]:
    """Acrescenta pares (parceiro, ação) pelo critério, sem violar nada.

    Não tira nem troca o que já está em `genes`: só preenche parceiros sem ação,
    enquanto couber no orçamento, em K e no máximo da categoria.
    """
    genes = list(genes)
    custo = sum(inst.custo[g - 1] for g in genes if g)
    acoes = sum(1 for g in genes if g)
    por_categoria = [0] * inst.categorias
    for i, g in enumerate(genes):
        if g and inst.categoria[i] >= 0:
            por_categoria[inst.categoria[i]] += 1

    pares = [
        (i, a)
        for i in range(inst.parceiros)
        if not genes[i]
        for a in range(inst.acoes)
        if inst.ganho[i][a] > 0
    ]

    # Razão comparada em inteiros, por produto cruzado: dividir daria `float`,
    # e o desempate mudaria conforme a linguagem arredonda. Um critério desempata
    # o outro; depois, menor parceiro e menor ação.
    def pela_razao(p: tuple[int, int], q: tuple[int, int]) -> int:
        (i, a), (j, b) = p, q
        esquerda, direita = inst.ganho[i][a] * inst.custo[b], inst.ganho[j][b] * inst.custo[a]
        return (esquerda < direita) - (esquerda > direita)

    def pelo_ganho(p: tuple[int, int], q: tuple[int, int]) -> int:
        (i, a), (j, b) = p, q
        return (inst.ganho[i][a] < inst.ganho[j][b]) - (inst.ganho[i][a] > inst.ganho[j][b])

    primeiro, segundo = (pela_razao, pelo_ganho) if criterio == RAZAO else (pelo_ganho, pela_razao)

    def comparar(p: tuple[int, int], q: tuple[int, int]) -> int:
        return primeiro(p, q) or segundo(p, q) or (p > q) - (p < q)

    for i, a in sorted(pares, key=cmp_to_key(comparar)):
        if genes[i] or acoes >= inst.maximo_acoes:
            continue
        if custo + inst.custo[a] > inst.orcamento:
            continue
        k = inst.categoria[i]
        if k >= 0 and por_categoria[k] >= inst.maximo_categoria[k]:
            continue
        genes[i] = a + 1
        custo += inst.custo[a]
        acoes += 1
        if k >= 0:
            por_categoria[k] += 1
    return genes


def guloso(inst: Instancia, criterio: str = RAZAO) -> list[int]:
    """O plano guloso completo; levanta `Inviavel` quando as cotas não cabem."""
    if criterio not in (RAZAO, GANHO):
        raise ValueError(f"Critério desconhecido: {criterio!r}")
    genes = [0] * inst.parceiros
    barata = inst.acao_mais_barata
    for i in conjunto_minimo(inst):
        genes[i] = barata + 1
    return completar(inst, genes, criterio)

File: nucleo/cpp/guloso.cpp
// Os dois planos gulosos — reproduz `gih_nucleo/guloso.py` (`completar`, `guloso`).
#include <algorithm>
#include <utility>

#include "nucleo.hpp"

namespace gih {
namespace {

// O produto `a·b` em 128 bits, sem `__int128`: o MSVC não tem, e a razão
// ganho/custo é comparada por produto cruzado. Com ganho e custo em centavos, o
// produto passa de 64 bits a partir de ~R$ 30 milhões × R$ 30 milhões — longe
// do uso, mas o Python não estoura nunca, e o porte precisa responder igual.
struct U128 {
    std::uint64_t alto;
    std::uint64_t baixo;
};

U128 multiplicar(std::uint64_t a, std::uint64_t b) {
    const std::uint64_t a0 = a & 0xFFFFFFFFULL, a1 = a >> 32;
    const std::uint64_t b0 = b & 0xFFFFFFFFULL, b1 = b >> 32;
    const std::uint64_t p00 = a0 * b0, p01 = a0 * b1, p10 = a1 * b0, p11 = a1 * b1;
    const std::uint64_t meio = (p00 >> 32) + (p01 & 0xFFFFFFFFULL) + (p10 & 0xFFFFFFFFULL);
    return {p11 + (p01 >> 32) + (p10 >> 32) + (meio >> 32), (meio << 32) | (p00 & 0xFFFFFFFFULL)};
}

// -1, 0 ou 1, como `(a·b > c·d)`… ordenado do maior para o menor.
int decrescente_por_produto(std::int64_t a, std::int64_t b, std::int64_t c, std::int64_t d) {
    const U128 x = multiplicar(static_cast<std::uint64_t>(a), static_cast<std::uint64_t>(b));
    const U128 y = multiplicar(static_cast<std::uint64_t>(c), static_cast<std::uint64_t>(d));
    if (x.alto != y.alto) return x.alto > y.alto ? -1 : 1;
    if (x.baixo != y.baixo) return x.baixo > y.baixo ? -1 : 1;
    return 0;
}

int decrescente(std::int64_t x, std::int64_t y) {
    if (x == y) return 0;
    return x > y ? -1 : 1;
}

std::vector<Gene> completar(const Instancia& inst, std::vector<Gene> genes, Criterio criterio) {
    std::int64_t custo = 0, acoes = 0;
    std::vector<std::int64_t> por_categoria(static_cast<std::size_t>(inst.categorias), 0);
    for (int i = 0; i < inst.parceiros; ++i) {
        if (!genes[i]) continue;
        custo += inst.custo[genes[i] - 1];
        acoes += 1;
        if (inst.categoria[i] >= 0) por_categoria[inst.categoria[i]] += 1;
    }

    std::vector<std::pair<int, int>> pares;
    for (int i = 0; i < inst.parceiros; ++i) {
        if (genes[i]) continue;
        for (int a = 0; a < inst.acoes; ++a) {
            if (inst.ganho_de(i, a) > 0) pares.emplace_back(i, a);
        }
    }

    // Um critério desempata o outro; depois, menor parceiro e menor ação — a
    // ordem é total, e por isso o `sort` chega à mesma sequência do `sorted`.
    auto pela_razao = [&](const std::pair<int, int>& p, const std::pair<int, int>& q) {
        return decrescente_por_produto(inst.ganho_de(p.first, p.second), inst.custo[q.second],
                                       inst.ganho_de(q.first, q.second), inst.custo[p.second]);
    };
    auto pelo_ganho = [&](const std::pair<int, int>& p, const std::pair<int, int>& q) {
        return decrescente(inst.ganho_de(p.first, p.second), inst.ganho_de(q.first, q.second));
    };
    std::sort(pares.begin(), pares.end(), [&](const auto& p, const auto& q) {
        int r = criterio == Criterio::Razao ? pela_razao(p, q) : pelo_ganho(p, q);
        if (r == 0) r = criterio == Criterio::Razao ? pelo_ganho(p, q) : pela_razao(p, q);
        if (r != 0) return r < 0;
        return p < q;
    });

    for (const auto& [i, a] : pares) {
        if (genes[i] || acoes >= inst.maximo_acoes) continue;
        if (custo + inst.custo[a] > inst.orcamento) continue;
        const int k = inst.categoria[i];
        if (k >= 0 && por_categoria[k] >= inst.maximo_categoria[k]) continue;
        genes[i] = static_cast<Gene>(a + 1);
        custo += inst.custo[a];
        acoes += 1;
        if (k >= 0) por_categoria[k] += 1;
    }
    return genes;
}

}  // namespace

std::vector<Gene> guloso(const Instancia& inst, Criterio criterio) {
    std::vector<int> minimo;
    Inviabilidade motivo;
    diagnosticar(inst, minimo, motivo);  // quem chama já conferiu que cabe
    std::vector<Gene> genes(static_cast<std::size_t>(inst.parceiros), 0);
    const Gene barata = static_cast<Gene>(inst.acao_mais_barata() + 1);
    for (int i : minimo) genes[i] = barata;
    return completar(inst, std::move(genes), criterio);
}

}  // namespace gih
